message/validation/consensus_validation.go

```diff
@@ -0,0 +1,361 @@
+		return 12 // TODO: consider calculating based on quick timeout and slow timeout
+	case spectypes.BNRoleProposer, spectypes.BNRoleSyncCommittee, spectypes.BNRoleSyncCommitteeContribution:
+		return 6
+	case spectypes.BNRoleValidatorRegistration:
+		return 0
+	default:
+		panic("unknown role")
+	}
+}
+
+func (mv *MessageValidator) currentEstimatedRound(sinceSlotStart time.Duration) specqbft.Round {
+	if currentQuickRound := specqbft.FirstRound + specqbft.Round(sinceSlotStart/roundtimer.QuickTimeout); currentQuickRound <= roundtimer.QuickTimeoutThreshold {
+		return currentQuickRound
+	}
+
+	sinceFirstSlowRound := sinceSlotStart - (time.Duration(roundtimer.QuickTimeoutThreshold) * roundtimer.QuickTimeout)
+	estimatedRound := roundtimer.QuickTimeoutThreshold + specqbft.FirstRound + specqbft.Round(sinceFirstSlowRound/roundtimer.SlowTimeout)
+	return estimatedRound
+}
+
+func (mv *MessageValidator) waitAfterSlotStart(role spectypes.BeaconRole) time.Duration {
+	switch role {
+	case spectypes.BNRoleSyncCommitteeContribution:
+		return mv.netCfg.Beacon.SlotDurationSec() // TODO: do we need this?
+	case spectypes.BNRoleAttester, spectypes.BNRoleAggregator, spectypes.BNRoleSyncCommittee, spectypes.BNRoleProposer, spectypes.BNRoleValidatorRegistration:
+		return 0
+	default:
+		panic("unknown role")
+	}
+}
+
+func (mv *MessageValidator) validRole(roleType spectypes.BeaconRole) bool {
+	switch roleType {
+	case spectypes.BNRoleAttester,
+		spectypes.BNRoleAggregator,
+		spectypes.BNRoleProposer,
+		spectypes.BNRoleSyncCommittee,
+		spectypes.BNRoleSyncCommitteeContribution,
+		spectypes.BNRoleValidatorRegistration:
+		return true
+	}
+	return false
+}
+
+func (mv *MessageValidator) validConsensusSigners(share *ssvtypes.SSVShare, m *specqbft.SignedMessage) error {
+	if len(m.Signers) == 0 {
+		return ErrNoSigners
+	}
+
+	if len(m.Signers) == 1 {
+		if m.Message.MsgType == specqbft.ProposalMsgType {
+			qbftState := &specqbft.State{
+				Height: m.Message.Height,
+				Share:  &share.Share,
+			}
+			leader := specqbft.RoundRobinProposer(qbftState, m.Message.Round)
+			if m.Signers[0] != leader {
+				err := ErrSignerNotLeader
+				err.got = m.Signers[0]
+				err.want = leader
+				return err
+			}
+		}
+	} else if m.Message.MsgType != specqbft.CommitMsgType {
+		e := ErrNonDecidedWithMultipleSigners
+		e.got = len(m.Signers)
+		return e
+	} else if !share.HasQuorum(len(m.Signers)) || len(m.Signers) > len(share.Committee) {
+		e := ErrWrongSignersLength
+		e.want = fmt.Sprintf("between %v and %v", share.Quorum, share.Committee)
+		e.got = len(m.Signers)
+		return e
+	}
+
+	if !slices.IsSorted(m.Signers) {
+		return ErrSignersNotSorted
+	}
+
+	seen := map[spectypes.OperatorID]struct{}{}
+	for _, signer := range m.Signers {
+		if err := mv.commonSignerValidation(signer, share); err != nil {
+			return err
+		}
+
+		if _, ok := seen[signer]; ok {
+			return ErrDuplicatedSigner
+		}
+		seen[signer] = struct{}{}
+	}
```

Comment:
since the slice is sorted, we can make this cheaper check without a map allocation:

```suggestion
	var prevSigner spectypes.OperatorID
	for _, signer := range m.Signers {
		if err := mv.commonSignerValidation(signer, share); err != nil {
			return err
		}
		if signer == prevSigner {
			return ErrDuplicatedSigner
		}
		prevSigner = signer
	}
```
(signer is already verified to not be 0 so don't need to check for it)